util/unionfind.cpp

```diff
@@ -0,0 +1,32 @@
+// Copyright (c) 2018-present The Alive2 Authors.
+// Distributed under the MIT license that can be found in the LICENSE file.
+
+#include "util/unionfind.h"
+#include <vector>
+
+using namespace std;
+
+namespace util {
+
+UnionFind::UnionFind(unsigned n) {
+  for (unsigned i = 0; i < n; ++i) {
+    id.push_back(i);
+  }
+}
+
+unsigned UnionFind::root(unsigned i) {
+  while (i != id[i]) i = id[i];
+  return i;
```

Comment:
Can we update id[i] to the new root?
This is a known technique that helps further flatten the tree ([link](https://en.wikipedia.org/wiki/Disjoint-set_data_structure#Finding_set_representatives)).
```
unsigned UnionFind::root(unsigned i0) {
   unsigned i = i0;
   while (i != id[i]) i = id[i];
   id[i0] = i;
   return i;
}
```

Reply:
To be precise, all nodes on the path from i0 to the root can be updated to have the found root.